[HLSL][RootSignature] Add metadata generation for descriptor tables

inbelic opened the pull request:
- prereq: Modify `RootSignatureAttr` to hold a reference to the owned declaration
- Define and implement `MetadataBuilder` in `HLSLRootSignature`
- Integrate and invoke the builder in `CGHLSLRuntime.cpp` to generate the Root Signature for any associated entry functions
- Add tests to demonstrate functionality in `RootSignature.hlsl`

Resolves https://github.com/llvm/llvm-project/issues/126584

Note: this is essentially just https://github.com/llvm/llvm-project/pull/125131 rebased onto the new approach of constructing a root signature decl, instead of holding the elements in `AdditionalMembers`.

inbelic committed:
add basic empty root signature

inbelic committed:
pass down the actual root elements

- test that we have the correct number of elements

inbelic committed:
introduce a MetadataBuilder to handle the construction of nodes

inbelic committed:
add support for DescriptorTableClauses

inbelic committed:
add support for associating DescriptorTables with their clauses

inbelic committed:
self-review: misc clean up

inbelic committed:
review: remove use of CHECK-DAG

inbelic committed:
pre-req: update attr to store reference to root signature decl

inbelic committed:
clang-format

inbelic committed:
self-review: remove commented header

inbelic commented on llvm/lib/Frontend/HLSL/HLSLRootSignature.cpp:
```diff
@@ -160,6 +163,75 @@ void dumpRootElements(raw_ostream &OS, ArrayRef<RootElement> Elements) {
   OS << "}";
 }
 
+static MDString *ClauseTypeToName(LLVMContext &Ctx, ClauseType Type) {
```
note: We can actually probably use the `<<` operator defined above.

joaosaffran commented on llvm/lib/Frontend/HLSL/HLSLRootSignature.cpp:
```diff
@@ -160,6 +163,75 @@ void dumpRootElements(raw_ostream &OS, ArrayRef<RootElement> Elements) {
+  // Remaining operands are references to the table's clauses. The in-memory
+  // representation of the Root Elements created from parsing will ensure that
+  // the previous N elements are the clauses for this table.
+  assert(Table.NumClauses <= GeneratedMetadata.size() &&
+         "Table expected all owned clauses to be generated already");
```
Not sure if that assert would really help, since it could easily be mislead by the generation of other parameters.

inbelic replied:
This was added as a (hopefully) better error message rather than the error that would be generated if `Generated.size() - Table.NumClauses` created an undefined behaviour iterator.

Do you have a suggestion for how it could be better worded, or, a better assert?

joaosaffran replied:
Not sure how complex/useful this is, we could check how many `DescriptorTableClause` are in `TableOperands` preceding the table. I think this could be done like:

```C++
uint Count = 0;
// I think this should be the Element parameter, from BuildRootSignature
auto It = Element;
while (std::holds_alternative<DescriptorTableClause>(*It)){
    Count ++;
    Element --;
}

assert(Table.NumClauses == Count);
```

Not sure this code is correct/useful, but that is the idea.

inbelic replied:
I like the idea and toyed around with it a bit. It unfortunately is a bit a unwieldly to implement. I think it introduces more complexity then it reduces

joaosaffran commented on llvm/lib/Frontend/HLSL/HLSLRootSignature.cpp:
```diff
@@ -160,6 +163,75 @@ void dumpRootElements(raw_ostream &OS, ArrayRef<RootElement> Elements) {
+  assert(Table.NumClauses <= GeneratedMetadata.size() &&
+         "Table expected all owned clauses to be generated already");
+  // So, add a refence to each clause to our operands
+  TableOperands.append(GeneratedMetadata.end() - Table.NumClauses,
+                       GeneratedMetadata.end());
```
So we ensure those ranges are being written in the correct order?

inbelic replied:
Correct, I added a comment to documentation the in-memory representation better

inbelic committed:
self-review: use << operator instead of redefining switch behaviour

inbelic committed:
self-review: mark to revisit get_if chain

inbelic committed:
review: clarify the struct the in-memory representation

- formally document the structure of the in-memory representation of a
root signature as an array of the defined RootElements

bogner commented on clang/lib/CodeGen/CGHLSLRuntime.cpp:
```diff
@@ -423,6 +437,14 @@ void CGHLSLRuntime::emitEntryFunction(const FunctionDecl *FD,
   B.CreateRetVoid();
+
+  // Add and identify root signature to function, if applicable
+  const AttrVec &Attrs = FD->getAttrs();
+  for (const Attr *Attr : Attrs) {
```
I don't think naming `Attrs` is doing much here, probably simpler just to call the function in the loop header
```suggestion
  for (const Attr *Attr : FD->getAttrs()) {
```

bogner commented on llvm/lib/Frontend/HLSL/HLSLRootSignature.cpp:
```diff
@@ -160,6 +163,65 @@ void dumpRootElements(raw_ostream &OS, ArrayRef<RootElement> Elements) {
+}
+
+MDNode *MetadataBuilder::BuildDescriptorTable(const DescriptorTable &Table) {
+  IRBuilder<> B(Ctx);
```
`B` is not a good name for an `IRBuilder` (despite that it does seem to be used a fair amount). `IRB` and `Builder` are both more common names for these, and I definitely lean towards the latter.

bogner reviewed (approved):
Looks good. A couple of style nitpicks.

inbelic committed:
review: style fixups
